sdk/core/azure-core/src/main/java/com/azure/core/http/HttpClient.java

```diff
@@ -31,6 +31,17 @@ default Mono<HttpResponse> send(HttpRequest request, Context context) {
         return send(request);
     }
 
+    /**
+     * Sends the provided request synchronously with contextual information.
+     *
+     * @param request The HTTP request to send.
+     * @param context Contextual information about the request.
+     * @return The response.
+     */
+    default HttpResponse sendSync(HttpRequest request, Context context) {
+        return send(request, context).block();
+    }
```

Comment:
I wonder if we should eagerly add `Duration timeout` parameter to `sendSync` in each layer.

The reason is that APIs at sync clients take that as parameter and block on reactor like this:
https://github.com/Azure/azure-sdk-for-java/blob/e849033d7fe961b0e3dd71d9eae44d9e712de564/sdk/storage/azure-storage-blob/src/main/java/com/azure/storage/blob/BlobClient.java#L298-L310

Netty and OkHttp client can implement "call timeout" (i.e. timeout around whole exchange). With netty we'd have to block here in default impl, then with OkHttp we'd use `callTimeout` . JDK client' isn't great in this case https://bugs.openjdk.java.net/browse/JDK-8258397 .

As for actual implementation of this I can see three options:
1. Like I said above. We can add `Duration timeout` as explicit param `sendSync`
2. We could add optional `HttpRequest` property.
3. We could leverage `Context` to smuggle it like here https://github.com/Azure/azure-sdk-for-java/blob/e849033d7fe961b0e3dd71d9eae44d9e712de564/sdk/core/azure-core-http-netty/src/main/java/com/azure/core/http/netty/NettyAsyncHttpClient.java#L111 .

I'd suggest we do 1 or 2. @alzimmermsft @g2vinay @srnagar @JonathanGiles  please share your thoughts.

Reply:
I think eagerly getting the right API at the lower layers is a good idea. We'll be setting ourselves up for success if we do that, rather than leaving landmines.

However, we don't strongly specify in the guidelines (as far as I can see or recall) a requirement around having sync methods taking a timeout. We should rectify this to be clear around our expectations for user-facing API. Assuming that we specify that all user-facing sync methods must have some mechanism for providing a timeout, we then need to consider how we do that, which is essentially what you've said above, except from the perspective of the user-facing API:

1. All methods could have a `Duration timeout` parameter, but this increases the barrier to usage, as it requires developers to determine what timeout to specify (and much of the time they would not care, so it is likely to be null, which decreases future readability).
2. We could require that for the max `withResponse` overload that we have both `Duration timeout` and `Context context` as required parameters. This minimises the frequency with which developers will encounter a `Duration timeout` parameter to just the maximal overloads, but it also makes proper usage of the sync APIs (with a timeout specified) less user-friendly as they will be encountering the maximal overloads).
3. We introduce some convenience API around `Context`, e.g. `Contexts.create().timeout(..).foo(..)` that we can train users to use to more easily smuggle common things into the Context.

Reply:
From what I can find there are three services that take timeout in sync apis - storage, search and tables. 
From what I recall the timeout parameter was there as a result of lack of default timeouts/not great story around http client timeouts. The number 3 is what I'd do for user facing APIs given that we'd rather direct them to HttpClientOptions for fine tuning (except couple of users who need per api call control). 

If we introduce `Context.timeout()` I'm fine with that approach (also for lower layers - mostly because http policy authors would just be passing it down do http client - easy to make mistake and drop this parameter). I'm trying to avoid introducing it as a "special key" in the context.

Reply:
The way how this is implemented (see snippet in the original comment) is that sync client blocks with timeout on async client equivalent. So effectively all these timeouts are "per entire API call".
Therefore in lower layers (retries mostly) we'd need to do the math to preserve semantics. E.g, keep a stopwatch in the request context and pass remaining duration to subsequent retires.

Reply:
This work is tracked in a separate issue - https://github.com/Azure/azure-sdk-for-java/issues/29156

Reply:
Fwiw, a small nuance to what I said above, I used the name `Contexts` (plural) when talking about convenience API on Context, rather than adding convenience API on `Context` directly. I'm not strongly wedded to it, but I wanted to make the distinction clear.

Reply:
@JonathanGiles what's return type of `Contexts.create()` in `Contexts.create().timeout(..).foo(..)` above ?

Reply:
Eventually it would be a `Context`

Reply:
Here's more about `Contexts` https://github.com/Azure/azure-sdk-for-java/pull/29277#discussion_r896214171 . It will be done separately from this PR.